conseil-common/src/main/scala/tech/cryptonomic/conseil/common/rpc/RpcClient.scala

```diff
@@ -0,0 +1,176 @@
+      encode: EntityEncoder[F, List[RpcRequest[P]]]
+  ): Stream[F, R] =
+    request
+      .chunkN(batchSize)
+      .evalTap(requests => debug(s"Call RPC in a batch of: ${requests.size}"))
+      // From JSON-RPC documentation: The Server MAY process a batch rpc call as a set of concurrent tasks, 
+      //                              processing them in any order and with any width of parallelism.
+      // So we also can process the responses in unordered order.
+      .mapAsyncUnordered(maxConcurrent) { chunks =>
+        httpClient
+          .expect[Seq[RpcResponse[R]]](
+            Method.POST(chunks.toList, Uri.unsafeFromString(endpoint), headers: _*)
+          )
+      }
+      .flatMap(Stream.emits)
+      .flatMap {
+        case RpcResponse(_, _, Some(error), _) =>
+          Stream.raiseError[F](error)
+        case RpcResponse(_, Some(result), _, _) =>
+          Stream(result)
+        case response =>
+          Stream.raiseError[F](
+            new UnsupportedOperationException(
+              s"Unexpected response from JSON-RPC server at $endpoint $response"
+            )
+          )
+      }
+
+  /**
+    * Helper method to log debug information.
+    * It wraps scalalogging call into [[Concurrent]] to prevent side effects.
+    * It should be moved to separate logging util in the future.
+    *
+    * @param message Debug message
+    */
+  private def debug(message: String): F[Unit] = Concurrent[F].delay(logger.debug(message))
+}
+
+object RpcClient {
+
+  /**
+    * Create [[cats.Resource]] with [[RpcClient]].
+    *
+    * @param endpoint JSON-RPC server url
+    * @param maxConcurrent Max concurrent processes used to fetch data from JSON-RPC server
+    * @param httpClient Http4s http client
+    * @param headers Additional http headers, e.g. authorization
+    */
+  def resource[F[_]: Concurrent](
+      endpoint: String,
+      maxConcurrent: Int,
+      httpClient: Client[F],
+      headers: Header*
+  ): Resource[F, RpcClient[F]] =
+    for {
+      client <- Resource.pure(
+        new RpcClient[F](
+          endpoint,
+          maxConcurrent,
+          httpClient,
+          headers: _*
+        )
+      )
+    } yield client
+
+  /**
+    * Exception with JSON-RPC error message
+    *
+    * Defined Error Codes:
+    * -32700 ---> parse error. not well formed
+    * -32701 ---> parse error. unsupported encoding
+    * -32702 ---> parse error. invalid character for encoding
+    * -32600 ---> server error. invalid xml-rpc. not conforming to spec.
+    * -32601 ---> server error. requested method not found
+    * -32602 ---> server error. invalid method parameters
+    * -32603 ---> server error. internal xml-rpc error
+    * -32500 ---> application error
+    * -32400 ---> system error
+    * -32300 ---> transport error
+    *
+    * In addition, the range -32099 .. -32000, inclusive is reserved for implementation defined server errors.
+    */
+  case class RpcException(
```

Comment:
Would be nice to know what we can expect in `data` (more than what we can see in `message`) too.
Asking cause I'm not sure if writing this down in expiation message is good idea (if `data` is quite large then probably not).

Reply:
The `data` field is defined in the JSON-RPC https://www.jsonrpc.org/specification

> data
> A Primitive or Structured value that contains additional information about the error.
> This may be omitted.
> The value of this member is defined by the Server (e.g. detailed error information, nested errors etc.). 

I think that bitcoin doesn't use it, but the field should stay to comply with the specification.